idunn/api/places_list.py

```diff
@@ -152,86 +169,109 @@ def valid_categories(cls, v):
-            detail="Both 'raw_filter' and 'category' parameters cannot be provided together",
-        )
-    return raw_params
+class PlacesBboxResponse(BaseModel):
+    places: List[Any]
+    source: PoiSource
+    bbox: Optional[Tuple[float, float, float, float]] = Field(
+        description="Minimal bbox containing all results. `null` if no result is found. May be larger than or outside of the original bbox passed in the query if `?extend_bbox=true` was set.",
+        example=(2.32, 48.85, 2.367, 48.866),
+    )
+    bbox_extended: bool = Field(
+        description="`true` if `?extend_bbox=true` was set and search has been executed on an extended bbox, after no result was found in the original bbox passed in the query."
+    )
+
+    @validator("bbox")
+    def round_bbox_values(cls, v):
+        if v is None:
+            return v
+        return tuple(round(x, 6) for x in v)
 
 
 async def get_places_bbox(
-    bbox: Any,
+    bbox: str = Query(
+        ...,
+        title="Bounding box to search",
+        description="Format: left_lon,bottom_lat,right_lon,top_lat",
+        example="-4.56,48.35,-4.42,48.46",
+    ),
     category: Optional[List[str]] = Query(None),
     raw_filter: Optional[List[str]] = Query(None),
     source: Optional[str] = Query(None),
-    q: Optional[str] = Query(None),
+    q: Optional[str] = Query(None, title="Query", description="Full text query"),
     size: Optional[int] = Query(None),
     lang: Optional[str] = Query(None),
     verbosity: Optional[str] = Query(None),
-):
-    es = get_elasticsearch()
-    raw_params = get_raw_params(bbox, category, raw_filter, source, q, size, lang, verbosity)
-    try:
-        params = PlacesQueryParam(**raw_params)
-    except ValidationError as e:
-        logger.info(f"Validation Error: {e.json()}")
-        raise HTTPException(status_code=400, detail=e.errors())
-
+    extend_bbox: bool = Query(False),
+) -> PlacesBboxResponse:
+    params = PlacesQueryParam(**locals())
     source = params.source
     if source is None:
         if (
             params.q or (params.category and all(c.get("pj_filters") for c in params.category))
         ) and pj_source.bbox_is_covered(params.bbox):
-            source = SOURCE_PAGESJAUNES
+            params.source = PoiSource.PAGESJAUNES
         else:
-            source = SOURCE_OSM
+            params.source = PoiSource.OSM
+
+    places_list = await _fetch_places_list(params)
+    bbox_extended = False
+
+    if params.extend_bbox and len(places_list) == 0:
+        original_bbox = params.bbox
+        original_bbox_width = original_bbox[2] - original_bbox[0]
+        original_bbox_height = original_bbox[3] - original_bbox[1]
+        original_bbox_size = max(original_bbox_height, original_bbox_width)
+        if original_bbox_size < EXTENDED_BBOX_MAX_SIZE:
+            # Compute extended bbox and fetch results a second time
+            scale_factor = EXTENDED_BBOX_MAX_SIZE / original_bbox_size
```

Comment:
Don't we have a spatial library to do this kind of error-prone operations for us?

Reply:
You're right. Using [`shapely.affinity.scale`](https://shapely.readthedocs.io/en/latest/manual.html#shapely.affinity.scale) makes things a little simpler.